CromIAM/src/main/scala/cromiam/sam/SamClient.scala

```diff
@@ -35,49 +37,67 @@ class SamClient(scheme: String,
+      )
       whitelisted <- response.status match {
-        case StatusCodes.OK => Unmarshal(response.entity).to[String].map(_.toBoolean)
-        case _ => Future.successful(false)
+        case StatusCodes.OK =>
+          implicit val whyIsThisNotAutomaticAreThereImportsMissing: Unmarshaller[HttpEntity, Boolean] =
+            (Unmarshaller.stringUnmarshaller flatMap Unmarshaller.booleanFromStringUnmarshaller).asScala
+          val unmarshal = IO.fromFuture(IO(Unmarshal(response.entity).to[Boolean]))
+          FailureResponseOrT.right[HttpResponse](unmarshal)
+        case _ => FailureResponseOrT.right[HttpResponse](IO.pure(false))
       }
       _ = if (!whitelisted) log.error("Submit Access Denied for user {}", user.userId)
     } yield whitelisted
   }
 
-  def collectionsForUser(user: User, cromIamRequest: HttpRequest): Future[List[Collection]] = {
+  def collectionsForUser(user: User, cromIamRequest: HttpRequest): FailureResponseOrT[List[Collection]] = {
     val request = HttpRequest(method = HttpMethods.GET, uri = samBaseCollectionUri, headers = List[HttpHeader](user.authorization))
 
     for {
-      response <- instrumentRequest(() => Http().singleRequest(request), cromIamRequest, instrumentationPrefixForSam(userCollectionPrefix))
-      resources <- Unmarshal(response.entity).to[List[SamResource]]
+      response <- instrumentRequest(
+        () => Http().singleRequest(request).asFailureResponseOrT,
+        cromIamRequest,
+        instrumentationPrefixForSam(userCollectionPrefix)
+      )
+      futureIO = IO.fromFuture(IO(Unmarshal(response.entity).to[List[SamResource]]))
+      resources <- FailureResponseOrT.right(futureIO)
     } yield resources.map(r => Collection(r.resourceId))
   }
 
   /**
     * Requests authorization for the authenticated user to perform an action from the Sam service for a collection
     * @return Successful future if the auth is accepted, a Failure otherwise.
     */
-  def requestAuth(authorizationRequest: CollectionAuthorizationRequest, cromIamRequest: HttpRequest): Future[Unit] = {
+  def requestAuth(authorizationRequest: CollectionAuthorizationRequest,
+                  cromIamRequest: HttpRequest): FailureResponseOrT[Unit] = {
     val logString = authorizationRequest.action + " access for user " + authorizationRequest.user.userId +
       " on a request to " + authorizationRequest.action +  " for collection " + authorizationRequest.collection.name
 
-    def validateEntityBytes(byteString: ByteString): Future[Unit] = {
-      if (byteString.utf8String == "true") Future.successful(())
-      else {
+    def validateEntityBytes(byteString: ByteString): FailureResponseOrT[Unit] = {
+      if (byteString.utf8String == "true") {
+        FailureResponseOrT.right(IO.unit)
+      } else {
         log.warning("Sam denied " + logString)
-        Future.failed(new SamDenialException)
+        FailureResponseOrT.right(IO.raiseError[Unit](new SamDenialException))
```

Comment:
I don't really understand the meaning of raising an error in a `.right` (which notionally indicates success?)

Reply:
Changed these to the hopefully clearer `FailureResponseOrT[IO, HttpResponse, Unit](IO.raiseError(…`